reactos/drivers/usb/usbhub_new/ioctl.c

```diff
@@ -0,0 +1,1406 @@
+                       1,
+                       FALSE);
+
+    if (!InterlockedDecrement(&HubExtension->PendingRequestCount))
+    {
+        KeSetEvent(&HubExtension->PendingRequestEvent,
+                   EVENT_INCREMENT,
+                   FALSE);
+    }
+
+Exit:
+
+    PortExtension->PortPdoFlags &= ~USBHUB_PDO_FLAG_PORT_RESSETING;
+
+    USBH_CompleteIrp(WorkItemReset->Irp, Status);
+
+    WorkItemReset->PortExtension->PortPdoFlags &= ~USBHUB_PDO_FLAG_PORT_RESTORE_FAIL;
+}
+
+NTSTATUS
+NTAPI
+USBH_PdoIoctlResetPort(IN PUSBHUB_PORT_PDO_EXTENSION PortExtension,
+                       IN PIRP Irp)
+{
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    PUSBHUB_RESET_PORT_CONTEXT HubWorkItemBuffer;
+    PUSBHUB_IO_WORK_ITEM HubIoWorkItem;
+    NTSTATUS Status;
+
+    HubExtension = PortExtension->HubExtension;
+
+    DPRINT("USBH_PdoIoctlResetPort ... \n");
+
+    if (PortExtension->PortPdoFlags & USBHUB_PDO_FLAG_PORT_RESSETING)
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    Status = USBH_AllocateWorkItem(HubExtension,
+                                   &HubIoWorkItem,
+                                   USBH_ResetPortWorker,
+                                   sizeof(USBHUB_RESET_PORT_CONTEXT),
+                                   (PVOID *)&HubWorkItemBuffer,
+                                   DelayedWorkQueue);
+
+    if (!NT_SUCCESS(Status))
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        USBH_CompleteIrp(Irp, Status);
+        return Status;
+    }
+
+    PortExtension->PortPdoFlags |= USBHUB_PDO_FLAG_PORT_RESSETING;
+    IoMarkIrpPending(Irp);
+
+    HubWorkItemBuffer->PortExtension = PortExtension;
+    HubWorkItemBuffer->Irp = Irp;
+
+    Status = STATUS_PENDING;
+
+    USBH_QueueWorkItem(PortExtension->HubExtension, HubIoWorkItem);
+
+    return Status;
+}
+
+VOID
+NTAPI
+USBH_PortIdleNotificationCancelRoutine(IN PDEVICE_OBJECT Device,
+                                       IN PIRP Irp)
+{
+    PUSBHUB_PORT_PDO_EXTENSION PortExtension;
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    PIRP PendingIdleIrp = NULL;
+    PUSBHUB_IO_WORK_ITEM HubIoWorkItem;
+    PUSBHUB_IDLE_PORT_CANCEL_CONTEXT HubWorkItemBuffer;
+    NTSTATUS Status;
+
+    DPRINT("USBH_PortIdleNotificationCancelRoutine ... \n");
+
+    PortExtension = Device->DeviceExtension;
+    PortExtension->PortPdoFlags &= ~USBHUB_PDO_FLAG_IDLE_NOTIFICATION;
+
+    HubExtension = PortExtension->HubExtension;
+
+    PortExtension->IdleNotificationIrp = NULL;
+
+    if (HubExtension &&
+        HubExtension->HubFlags & USBHUB_FDO_FLAG_WAIT_IDLE_REQUEST)
+    {
+        PendingIdleIrp = HubExtension->PendingIdleIrp;
+        HubExtension->PendingIdleIrp = NULL;
+    }
+
+    IoReleaseCancelSpinLock(Irp->CancelIrql);
+
+    if (PendingIdleIrp)
+    {
+        USBH_HubCancelIdleIrp(HubExtension, PendingIdleIrp);
+    }
+
+    if (HubExtension->CurrentPowerState.DeviceState == PowerDeviceD0)
```

Comment:
ioctl.c:447: Above, in line 433 you check if HubExtension is NULL. If it can be NULL, then you will dereference a NULL pointer here. If it cannot be NULL, you can remove the check above.